### PR TITLE
Fix list extensions tree with multi level extensions

### PR DESCRIPTION
**Pull request by felixsch**

### CODE REVIEW

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/SCC-Connect-Prophet/92158/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project connect on step docker-run-tests.
( Failure: Merged e02d03357a42cbd63ec7f9202fe4b91a3c2feeb1 into 9b48c8d9dddcdee0f90e621c0f9fc92c20ced104 )

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/SCC-Connect-Prophet/92385/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project connect on step docker-run-tests.
Failed fast in project connect on step docker-run-tests.
( Failure: Merged 07553226bbcf6783d8dec287b84d7697bb9d7164 into 9b48c8d9dddcdee0f90e621c0f9fc92c20ced104 )

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/SCC-Connect-Prophet/92397/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
( Success: Merged dbeb8ed7a3ef83e8e29a8b82e2fbe9bedb7c0da6 into 9b48c8d9dddcdee0f90e621c0f9fc92c20ced104 )

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/SCC-Connect-Prophet/93251/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project connect on step docker-run-tests.
( Failure: Merged dbeb8ed7a3ef83e8e29a8b82e2fbe9bedb7c0da6 into 78a78f7f4a172a6285ab32f25cdc25d27d6cd6a9 )

File: lib/suse/connect/templates/extension.text.erb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<%= render 'extension_item.text', { extension: extension, any_installed: any_installed, level: level} %>
+<% if extension[:extensions] -%>
+<%   extension[:extensions].each do |sub_extension| -%>
```
Comment (lagartoflojo) [inline lib/suse/connect/templates/extension.text.erb]: You can remove the first `if extension[:extensions]` by changing the loop to: `extension.fetch(:extensions, []).each do |sub_extension|`

File: lib/suse/connect/templates/extension_item.text.erb [hunk 2]
```diff
@@ -1,11 +1,11 @@
-\e[1m<%= extension[:name] %>\e[0m<%
+<%= indent(level) %>\e[1m<%= extension[:name] %>\e[0m<%
   if extension[:installed]
     %> \e[32m(Installed)\e[0m
-  <%= '  ' if defined?(nested) %>De-Install with: <%= '  ' unless defined?(nested) %>SUSEConnect \e[31m-d\e[0m -p <%= extension[:activation_code] %><%
+<%= indent(level) %>De-Install with: SUSEConnect \e[31m-d\e[0m -p <%= extension[:activation_code] %><%
```
Comment (lagartoflojo) [inline lib/suse/connect/templates/extension_item.text.erb]: `De-Install` => `Deactivate`

File: lib/suse/connect/templates/extension_item.text.erb [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
   else %><%
     if extension[:activated]
       %> \e[33m(Activated)\e[0m<%
     end %>
-  <%= '  ' if defined?(nested) %>Install with:    <%= '  ' unless defined?(nested) %>SUSEConnect<%= '   ' if any_installed %> -p <%= extension[:activation_code] %><% unless extension[:free] %> -r \e[32m\e[1mREGCODE\e[0m<%
+<%= indent(level) %>Install with: SUSEConnect<%= '   ' if any_installed %> -p <%= extension[:activation_code] %><% unless extension[:free] %> -r \e[32m\e[1mREGCODE\e[0m<%
```
Comment (lagartoflojo) [inline lib/suse/connect/templates/extension_item.text.erb]: `Install` => `Activate`

This was not technically correct before, as we are activating the product, not installing it.

File: lib/suse/toolkit/renderer.rb [hunk 4]
```diff
@@ -7,11 +7,21 @@ def initialize(*arguments)
...
-      def render(filename, locals: binding)
-        bind = locals.is_a?(Binding) ? locals : OpenStruct.new(locals).instance_eval { binding }
-        @templates[filename] ||= ERB.new File.read(File.expand_path("../../connect/templates/#{filename}.erb", __FILE__)), 0, '-<>'
+      def render(filename, locals = {})
+        bind = binding
+        path = "../../connect/templates/#{filename}.erb"
+
+        locals.each_pair do |key, value|
+          bind.local_variable_set(key, value)
+        end
```
Comment (lagartoflojo) [inline lib/suse/toolkit/renderer.rb]: Why the change to the bindings logic?

File: lib/suse/toolkit/renderer.rb [hunk 5]
```diff
@@ -7,11 +7,21 @@ def initialize(*arguments)
...
+
+        @templates[filename] ||= ERB.new File.read(File.expand_path(path, __FILE__)), 0, '-<>'
```
Comment (lagartoflojo) [inline lib/suse/toolkit/renderer.rb]: Why this change?

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/SCC-Connect-Prophet/93415/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
( Success: Merged dbeb8ed7a3ef83e8e29a8b82e2fbe9bedb7c0da6 into 78a78f7f4a172a6285ab32f25cdc25d27d6cd6a9 )

File: lib/suse/toolkit/renderer.rb (hunk 4 above)
Comment (felixsch) [reply lib/suse/toolkit/renderer.rb]: I want that the renderer class methods are within the context (aka render and indent method). The old implementation does work for the first call where the binding from the default parameter is used. But in case when locals are supplied it evaluates the binding within the context of the openstruct which works fine but does not export the methods of renderer, resulting in calling 'render' or 'indent' is not known every render call except to the first one.

File: lib/suse/toolkit/renderer.rb (hunk 5 above)
Comment (felixsch) [reply lib/suse/toolkit/renderer.rb]: Yes this is my matter of taste. I basically only moved out the path as a own variable to make the line more short and more clear (at least for me). If you have objections I can also change it back.

File: lib/suse/connect/templates/extension.text.erb (hunk 1 above)
Comment (felixsch) [reply lib/suse/connect/templates/extension.text.erb]: Ah nice yes I do this

File: lib/suse/toolkit/renderer.rb (hunk 4 above)
Comment (lagartoflojo) [reply lib/suse/toolkit/renderer.rb]: Ok got it!

File: lib/suse/toolkit/renderer.rb (hunk 5 above)
Comment (lagartoflojo) [reply lib/suse/toolkit/renderer.rb]: I see, actually my comment was more about the `, 0, '-<>'` at the end. What's that about?

Comment (felixsch) [reply lib/suse/toolkit/renderer.rb]: So after checking out the documentation: the zero is the safe_level which is actually saying that erb is rendering in a separate thread.
For the '-<>': It defines how erb should handle blank lines and spaces.
http://ruby-doc.org/stdlib-2.5.0/libdoc/erb/rdoc/ERB.html#method-c-new

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/SCC-Connect-Prophet/93510/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
( Success: Merged e6d4977a6c23e5c114617dc7ab36f798d3b44e08 into 78a78f7f4a172a6285ab32f25cdc25d27d6cd6a9 )

File: lib/suse/toolkit/renderer.rb (hunk 5 above)
Comment (lagartoflojo) [reply lib/suse/toolkit/renderer.rb]: Ok, but I don't get it. We didn't have those before. Why do we need them now?

**Comment by lagartoflojo** [comment]
@felixsch actually your last commit https://github.com/SUSE/connect/pull/338/commits/3d9c1c68b8bfc6309b71ed272bf987b9e42a52e6 will break now for SLES 12....
Basically we need our tests to support both SLES 12 and SLES 15.

But don't worry about it, I'm working on that here https://github.com/SUSE/connect/pull/340.
Please revert your last commit, otherwise we'll have conflicts :)

**Comment by felixsch** [comment]
Ahhhh now I got it. Yes i revert the changes

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/SCC-Connect-Prophet/94037/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project connect on step docker-run-tests.
( Failure: Merged 3d9c1c68b8bfc6309b71ed272bf987b9e42a52e6 into 78a78f7f4a172a6285ab32f25cdc25d27d6cd6a9 )

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/SCC-Connect-Prophet/94038/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
( Success: Merged e6d4977a6c23e5c114617dc7ab36f798d3b44e08 into 78a78f7f4a172a6285ab32f25cdc25d27d6cd6a9 )

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/SCC-Connect-Prophet/94107/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
( Success: Merged 97075a1dacc58e1a59db4e970c4b327597842d8b into 78a78f7f4a172a6285ab32f25cdc25d27d6cd6a9 )

File: lib/suse/connect/templates/extension_item.text.erb [hunk 6]
```diff
@@ -1,11 +1,11 @@
...
   if extension[:installed]
     %> \e[32m(Installed)\e[0m
-  <%= '  ' if defined?(nested) %>De-Install with: <%= '  ' unless defined?(nested) %>SUSEConnect \e[31m-d\e[0m -p <%= extension[:activation_code] %><%
+<%= indent(level) %>Deactivate with: SUSEConnect \e[31m-d\e[0m -p <%= extension[:activation_code] %><%
   else %><%
     if extension[:activated]
       %> \e[33m(Activated)\e[0m<%
     end %>
-  <%= '  ' if defined?(nested) %>Install with:    <%= '  ' unless defined?(nested) %>SUSEConnect<%= '   ' if any_installed %> -p <%= extension[:activation_code] %><% unless extension[:free] %> -r \e[32m\e[1mREGCODE\e[0m<%
+<%= indent(level) %>Activate with: SUSEConnect<%= '   ' if any_installed %> -p <%= extension[:activation_code] %><% unless extension[:free] %> -r \e[32m\e[1mADDITIONAL REGCODE\e[0m<%
```
Comment (lagartoflojo) [inline lib/suse/connect/templates/extension_item.text.erb]: I'm not too hot about this "additional" regcode -- I would just say "regcode" as it was before.

**Comment by lagartoflojo** [comment]
The output on a SLES12 machine looks like this:

image

While it's fine, I would sort the modules/extensions by name. It's just a completely random list right now.

File: lib/suse/toolkit/renderer.rb (hunk 5 above)
Comment (lagartoflojo) [reply lib/suse/toolkit/renderer.rb]: I didn't see that they were already there before -- my mistake!

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/SCC-Connect-Prophet/95206/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
( Success: Merged 97075a1dacc58e1a59db4e970c4b327597842d8b into c98237c99982568f7e7e6809ca64c980d5266414 )

**Comment by felixsch** [comment]
Yes good idea.

**Comment by lagartoflojo** [comment]
Running connect on all 12 SPs and on 15 takes forever! I'm closing this so that https://github.com/SUSE/connect/pull/342 passes and I can merge it. Please re-open afterwards! :)

**Comment by suse-tests-fail** [comment]
Prophet reports failure.
https://ci.suse.de/job/SCC-Connect-Prophet/95411/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project connect on step docker-run-tests-sle12sp1. (Merged c22f1b25aa99de709243860f22a0cc8dc950f9fb into c98237c99982568f7e7e6809ca64c980d5266414)

**Comment by suse-tests-pass** [comment]
Well Done! Your tests are still passing.
https://ci.suse.de/job/SCC-Connect-Prophet/95440/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 7de239a9d75a9274c1d95b2cc331697b7a68bc84 into 05dddb0eca174d132e15114b068c5fba04387ec5)
